internal/datastore/crdb/caveat.go

```diff
@@ -2,24 +2,183 @@ package crdb
+		if err != nil {
+			return nil, fmt.Errorf("unable to list caveat: %w", err)
+		}
+		defer rows.Close()
+
+		var overlappingKeys []string
+		for rows.Next() {
+			var defBytes []byte
+			var overlappingKey string
+			err = rows.Scan(&overlappingKey, &defBytes)
+			if err != nil {
+				return nil, fmt.Errorf("unable to list caveat: %w", err)
+			}
+			allDefinitionBytes = append(allDefinitionBytes, defBytes)
+			overlappingKeys = append(overlappingKeys, overlappingKey)
+		}
+		return overlappingKeys, nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("unable to list caveat: %w", err)
+	}
+
+	caveats := make([]*core.CaveatDefinition, 0, len(allDefinitionBytes))
+	for _, defBytes := range allDefinitionBytes {
+		loaded := &core.CaveatDefinition{}
+		if err := loaded.UnmarshalVT(defBytes); err != nil {
+			return nil, fmt.Errorf("unable to list caveat: %w", err)
+		}
+		caveats = append(caveats, loaded)
+	}
+
+	return caveats, nil
 }
 
 func (rwt *crdbReadWriteTXN) WriteCaveats(caveats []*core.CaveatDefinition) error {
-	return fmt.Errorf("unimplemented caveat support in datastore")
+	ctx, span := tracer.Start(datastore.SeparateContextWithTracing(rwt.ctx), "WriteCaveats")
+	defer span.End()
+
+	write := writeCaveat
+	writtenCaveatNames := make([]string, 0, len(caveats))
+	for _, caveat := range caveats {
+		definitionBytes, err := caveat.MarshalVT()
+		if err != nil {
+			return err
+		}
+		valuesToWrite := []any{caveat.Name, definitionBytes}
+		write = write.Values(valuesToWrite...)
+		writtenCaveatNames = append(writtenCaveatNames, caveat.Name)
+	}
+	span.SetAttributes(common.CaveatNameKey.StringSlice(writtenCaveatNames))
+
+	// store the new caveat
+	sql, args, err := write.ToSql()
+	if err != nil {
+		return fmt.Errorf("unable to write new caveat revision: %w", err)
+	}
+
+	return rwt.executeWithOverlappingKeys(ctx, func(ctx context.Context, tx pgx.Tx) ([]string, error) {
+		if _, err := rwt.tx.Exec(ctx, sql, args...); err != nil {
+			return nil, fmt.Errorf("unable to write new caveat revision: %w", err)
+		}
+		return writtenCaveatNames, nil
+	})
 }
 
 func (rwt *crdbReadWriteTXN) DeleteCaveats(names []string) error {
-	return fmt.Errorf("unimplemented caveat support in datastore")
+	ctx, span := tracer.Start(datastore.SeparateContextWithTracing(rwt.ctx), "DeleteCaveats",
+		trace.WithAttributes(attribute.StringSlice("names", names)))
+	defer span.End()
+
+	deleteCaveatClause := deleteCaveat.Where(sq.Eq{colCaveatName: names})
+	sql, args, err := deleteCaveatClause.ToSql()
+	if err != nil {
+		return fmt.Errorf("unable to delete caveats: %w", err)
+	}
+	return rwt.executeWithOverlappingKeys(ctx, func(ctx context.Context, tx pgx.Tx) ([]string, error) {
+		if _, err := tx.Exec(ctx, sql, args...); err != nil {
+			return nil, fmt.Errorf("unable to delete caveats: %w", err)
+		}
+		return names, nil
+	})
+}
+
+func (cr *crdbReader) executeWithOverlappingKeys(ctx context.Context, f func(ctx context.Context, tx pgx.Tx) ([]string, error)) error {
```

Comment:
this is allocating an extra slice to track overlap keys, why not just loop over the names similar to how we do for namespaces?

there will almost always be only 1 key (and the "keyset" only stores it once), so it's not much help to store it N times. The default is even simpler - we discard the keys entirely (just write to a single static key instead).

Reply:
That is correct, it's a tradeoff I deliberately took, although I'm not 100% sold given the overhead. The goal was to spare us a bunch of repeated lines of code. This the `executeWithOverlappingKeys` could be used in other methods of the readerwriter implementation.

I'm not sure I follow what you mean the keyset stores it just once. Isn't each caveat name added to the keyset to compute the overlapping?
- `WriteCaveats` accepts a slice of caveats. Each name is an overlapping key.
- `DeleteCaveats` also accepts a slice of caveats, same here

Reply:
> I'm not sure I follow what you mean the keyset stores it just once. Isn't each caveat name added to the keyset to compute the overlapping?

There's two types keyset: `static` and `prefix`.

If it's a static keyset, then `addOverlapKey` is a no-op, we always touch the same key.

If it's a prefix keyset, then `addOverlapKey` extracts the prefix (i.e. `mynamespace/<whatever>`) and only stores the prefix.

In theory you can make requests with multiple prefixes, in practice it's always just 1 prefix.

By collecting up the keys like this, you'll build a big list in memory only to throw it away (because you just needed this single string prefix, not an array of all of the caveat names)

Reply:
@ecordell I've updated the code to handle overlapping key one level above and avoid the extra allocation

>In theory you can make requests with multiple prefixes, in practice it's always just 1 prefix.

I'd like to understand this better. Given `WriteCaveats` will be handled by API `WriteSchema`, wouldn't it be possible to send a schema write request with caveats / namespaces with different prefixes? If so, doesn't that require us to loop over all namespaces and caveats to make sure we don't miss any overlapping key?

Reply:
> If so, doesn't that require us to loop over all namespaces and caveats to make sure we don't miss any overlapping key?

Yes, and the "keyer" is for tracking all prefixes from the request. But most deployments will either not use prefixes at all (authzed dedicated, spicedb users) or will have a single prefix enforced a layer up (authzed serverless). And in SpiceDB deployments that do use prefixes, the default "overlap strategy" is `static` (meaning all namespaces write to a single static key).

If you use the "prefix" overlap strategy, then prefixes in SpiceDB w/ CRDB are really "new enemy protection domains" rather than just for organization. I don't think we have any telemetry on it, but I doubt there are many (if any) SpiceDB users who have bothered - it's only really useful if you are trying to enforce multitenancy on top of SpiceDB and/or you're hitting QPS limits with `static` (and can untangle the namespaces in some reasonable way / don't care about newenemy problem)

Reply:
Thanks for the context, helps a lot! Intentionally using prefix separation to reduce the bottleneck caused by serialization errors could indeed be an option for certain SpiceDB deployments, but still to be seen in real-life (beyond Authzed Serverless 😄 ). In any case, my take away is that, while it is relatively rare, we still need to make sure we account for all potential prefixes in an incoming request.